Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -235,22 +220,20 @@ bool NetworkStorageManager::canHandleTypes(OptionSet<WebsiteDataType> types)
         || types.contains(WebsiteDataType::IndexedDBDatabases);
 }
 
-void NetworkStorageManager::close()
+void NetworkStorageManager::close(CompletionHandler<void()>&& completionHandler)
 {
     ASSERT(RunLoop::isMain());
+    ASSERT(!m_closed);
 
-    if (m_closed)
-        return;
     m_closed = true;
-
     for (auto& connection : m_connections)
         connection.removeWorkQueueMessageReceiver(Messages::NetworkStorageManager::messageReceiverName());
 
-    m_queue->dispatch([this, protectedThis = Ref { *this }]() mutable {
+    m_queue->dispatch([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)]() mutable {
         m_originStorageManagers.clear();
         m_fileSystemStorageHandleRegistry = nullptr;
 
-        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis)] { });
+        RunLoop::main().dispatch(WTFMove(completionHandler));
```

Comment:
Why is it OK to no longer dispatch protectedThis to the main thread? How do we ensure that protectedThis here is not the last reference to the storage manager? Since we have 2 threads involved here, I don't think it is safe even if the completion handler does drop a ref to the NetworkStorageManager on the main thread.

Reply:
Okay, adding the `protectedThis` back.